File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        suffix = nums.copy()
        prefix = nums
        
        pre = 1
        for i in range(len(prefix)):
            temp = prefix[i]
            prefix[i] = pre
            pre = temp * pre
            print(temp)
            print(pre)
            print("num ", prefix[i])
        print(prefix)

        suf = 1
        for i in range(len(suffix) - 1, -1, -1):
            temp = suffix[i]
            suffix[i] = suf
            suf = temp * suf
        print(suffix)

        for i in range(0, len(prefix), 1):
            prefix[i] = prefix[i] * suffix[i]
        
        return prefix

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        if(nums[0] <= nums[nums.size()/2] && nums[nums.size()/2] <= nums[nums.size()-1])
         return nums[0];
        return nums[fmHelper(nums,0,nums.size()-1)];
    }
    int fmHelper(vector<int> &nums, int l, int r){
        if(l == r)
            return l;
        int mid = (l + r) /2;
        if(mid < nums.size()-1 && mid > 0 && nums[mid] < nums[mid - 1])
            return mid;
        if(nums[mid] < nums[r])
            return fmHelper(nums,l,mid -1);
        return fmHelper(nums,mid+1,r);
    }
};
